### PR TITLE
[TOOLS-2825] Timeline status calculation using scenario simulation

### PR DESCRIPTION
**Pull request by herbert-venancio**

### CODE REVIEW

**Comment by objective-solutions-user** [comment]
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: dashboard-scenario/src/main/java/objective/taskboard/scenario/ScenarioComponent.java [hunk 1]
```diff
@@ -37,25 +40,47 @@
...
+        this.services = new Services(scenarioService, clock);
+    }
+
+    public ProjectModel getModel(String projectKey) {
+        return getModel(projectKey, ZoneId.systemDefault());
+    }
+
+    public ProjectModel getModel(String projectKey, ZoneId timezone) {
+        ProjectFilterConfiguration project = projectService.getTaskboardProjectOrCry(projectKey);
+        return new ProjectModel(services, projectKey, project, timezone);
```
[the remark this answers is not in the document]
Comment (herbert-venancio) [reply dashboard-scenario/src/main/java/objective/taskboard/scenario/ScenarioComponent.java]: Ok

File: dashboard-scenario/src/test/java/objective/taskboard/scenario/ScenarioFacadeTest.java [hunk 2]
```diff
@@ -195,18 +199,6 @@ public void givenNoScenarios_whenListScenarios_fetchDefaultAverageScenario() {
...
-    @Test
-    public void givenNoScenarios_whenGetScenarioDetailsOfDefaultScenario_thenCreateFromDefaultProperties() {
-        given.todayIs(DEFAULT_2017_6_1);
-        given.noScenario();
-        given.randomDefaultProperties();
-
-        when.getDetailsOfDefaultScenario();
-
-        then.details()
-                .hasDefaultParameters();
```
[the remark this answers is not in the document]
Comment (herbert-venancio) [reply dashboard-scenario/src/test/java/objective/taskboard/scenario/ScenarioFacadeTest.java]: Ok

File: dashboard-scenario/src/main/java/objective/taskboard/scenario/model/ProjectModel.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        this.services = services;
+        this.projectKey = projectKey;
+        this.entity = entity;
+        this.clock = new ClockModel(services, timezone);
+        this.scopeBacklogSupplier = scopeBacklogSupplier;
+    }
+
+    private Scenario getScenario(long scenarioId) {
+        if(scenarioId == 0L)
+            return null;
```
[the remark this answers is not in the document]
Comment (herbert-venancio) [reply dashboard-scenario/src/main/java/objective/taskboard/scenario/model/ProjectModel.java]: Ok

File: dashboard-scenario/src/main/java/objective/taskboard/scenario/model/ScenarioModel.java [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+    }
+
+    public LocalDate calculateConservativeDeliveryDate() {
+        BigDecimal scopeConservative = parameters().forCurrentWeek().getScopeConservative();
+        BigDecimal scopeConservativeFactor = BigDecimal.ONE.add(scopeConservative);
+        BigDecimal projectDuration = BigDecimal.valueOf(project.durationInDays());
+        return project.startDate().plusDays(projectDuration.multiply(scopeConservativeFactor).intValue());
+    }
+
+    public BigDecimal calculateOptimisticWeeklyThroughput(long week) {
```
[the remark this answers is not in the document]
Comment (herbert-venancio) [reply dashboard-scenario/src/main/java/objective/taskboard/scenario/model/ScenarioModel.java]: Ok

File: dashboard-scenario/src/main/java/objective/taskboard/scenario/model/ScenarioModel.java [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+        WeekParametersModel fetchWeekParameters(long week, LocalDate weekDate) {
+            throw new NotImplementedException("Average Forecast not implemented yet");
+        }
+    }
+
+    class ScenarioWeekParameters extends WeekParametersSource {
+
+        @Override
+        WeekParametersModel fetchWeekParameters(long week, LocalDate weekDate) {
+            // perfect match
```
[the remark this answers is not in the document]
Comment (herbert-venancio) [reply dashboard-scenario/src/main/java/objective/taskboard/scenario/model/ScenarioModel.java]: Ok

File: dashboard-scenario/src/main/java/objective/taskboard/scenario/model/ScenarioModel.java [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+
+    private LocalDate calculateOptimisticDeliveryDate() {
+        BigDecimal scopeOptimistic = parameters().forCurrentWeek().getScopeOptimistic();
+        BigDecimal scopeOptimisticFactor = BigDecimal.ONE.subtract(scopeOptimistic);
+        BigDecimal projectDuration = BigDecimal.valueOf(project.durationInDays());
+        return project.startDate().plusDays(projectDuration.multiply(scopeOptimisticFactor).intValue());
+    }
+
+    public LocalDate calculateConservativeDeliveryDate() {
+        BigDecimal scopeConservative = parameters().forCurrentWeek().getScopeConservative();
```
[the remark this answers is not in the document]
Comment (herbert-venancio) [reply dashboard-scenario/src/main/java/objective/taskboard/scenario/model/ScenarioModel.java]: Ok

File: dashboard-scenario/src/main/java/objective/taskboard/scenario/model/ScenarioModel.java [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+    }
+
+    private LocalDate calculateOptimisticDeliveryDate() {
+        BigDecimal scopeOptimistic = parameters().forCurrentWeek().getScopeOptimistic();
+        BigDecimal scopeOptimisticFactor = BigDecimal.ONE.subtract(scopeOptimistic);
+        BigDecimal projectDuration = BigDecimal.valueOf(project.durationInDays());
+        return project.startDate().plusDays(projectDuration.multiply(scopeOptimisticFactor).intValue());
+    }
+
+    public LocalDate calculateConservativeDeliveryDate() {
```
[the remark this answers is not in the document]
Comment (herbert-venancio) [reply dashboard-scenario/src/main/java/objective/taskboard/scenario/model/ScenarioModel.java]: Ok

File: dashboard-scenario/src/main/java/objective/taskboard/scenario/model/ScenarioModel.java [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+        return parameters.get(() -> {
+            Scenario scenario = entity();
+            if(scenario == null)
+                return new AverageForecastWeekParameters();
+
+            return new ScenarioWeekParameters();
+        });
+    }
+
+    private LocalDate calculateOptimisticDeliveryDate() {
```
[the remark this answers is not in the document]
Comment (herbert-venancio) [reply dashboard-scenario/src/main/java/objective/taskboard/scenario/model/ScenarioModel.java]: Ok

File: dashboard-scenario/src/main/java/objective/taskboard/scenario/model/ScenarioModel.java [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+        BigDecimal scopeOptimistic = parameters().forCurrentWeek().getScopeOptimistic();
```
[the remark this answers is not in the document]
Comment (herbert-venancio) [reply dashboard-scenario/src/main/java/objective/taskboard/scenario/model/ScenarioModel.java]: Ok

File: strategical-dashboard/src/test/java/objective/taskboard/monitor/TimelineMonitorCalculatorTest.java [hunk 10]
```diff
@@ -1,106 +1,345 @@
...
+        given(project()
+                .withProjectKey("PROJ")
+                .withStartDate(startDate)
+                .withDeliveryDate(startDate.plusDays(100)));
+        given.totalScopeIs(2000);
+
+        // scope delivered is following linear progress
+        given.todayIs(startDate.plusDays(50)).and.backlogIs(1000);
+
+        given(parameters()
```
[the remark this answers is not in the document]
Comment (herbert-venancio) [reply strategical-dashboard/src/test/java/objective/taskboard/monitor/TimelineMonitorCalculatorTest.java]: Ok

File: dashboard-scenario/src/main/java/objective/taskboard/scenario/model/ScenarioModel.java [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+                if(weekDate.isBefore(project.clock.currentWeek()))
+                    throw new IllegalArgumentException("Week parameters for week " + week + " (" + date + ") is before current week");
+                return fetchWeekParameters(week, weekDate);
+            });
+        }
+
+        abstract WeekParametersModel fetchWeekParameters(long week, LocalDate weekDate);
+    }
+
+    class AverageForecastWeekParameters extends WeekParametersSource {
```
[the remark this answers is not in the document]
Comment (herbert-venancio) [reply dashboard-scenario/src/main/java/objective/taskboard/scenario/model/ScenarioModel.java]: Ok

File: dashboard-scenario/src/main/java/objective/taskboard/scenario/model/ScenarioModel.java [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+
+    public ScenarioModel(ProjectModel project, Supplier<Scenario> scenarioSupplier) {
+        this.project = project;
+        this.entity = memoize(scenarioSupplier::get);
+    }
+
+    private WeekParametersSource parameters() {
+        return parameters.get(() -> {
+            Scenario scenario = entity();
+            if(scenario == null)
```
[the remark this answers is not in the document]
Comment (herbert-venancio) [reply dashboard-scenario/src/main/java/objective/taskboard/scenario/model/ScenarioModel.java]: Ok

File: dashboard-scenario/src/main/java/objective/taskboard/scenario/ScenarioService.java [hunk 13]
```diff
@@ -53,47 +53,33 @@ public synchronized ProjectScenario getOrCreateForProject(ProjectFilterConfigura
...
-                            .filter(entry -> entry.getKey().isBefore(week))
-                            .max(Comparator.comparing(Map.Entry::getKey))
-                            .map(Map.Entry::getValue);
-                })
-                .orElseGet(this::getOrCreateDefaultWeekParameters);
+        Scenario activeScenario = projectScenario.getActiveScenario();
+        if(activeScenario == null)
+            throw new NotImplementedException("Average Forecast not implemented yet");
+
+        WeekParameters weekParameters = activeScenario.getWeekParameters().entrySet().stream()
```
[the remark this answers is not in the document]
Comment (herbert-venancio) [reply dashboard-scenario/src/main/java/objective/taskboard/scenario/ScenarioService.java]: Ok

File: dashboard-scenario/src/main/java/objective/taskboard/scenario/model/ScenarioModel.java [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+                        simulateDeliveryDate(this::calculateOptimisticWeeklyThroughput)
+                        , simulateDeliveryDate(this::calculateConservativeWeeklyThroughput)
+                ));
+            } catch (IllegalStateException ex) { // NOSONAR
+                return Optional.empty();
+            }
+        });
+    }
+
+    public Optional<BigDecimal> simulatedToPlannedIntersectionRate() {
```
[the remark this answers is not in the document]
Comment (herbert-venancio) [reply dashboard-scenario/src/main/java/objective/taskboard/scenario/model/ScenarioModel.java]: Ok

**Pull request merged**